lib/endpoints/vnc.js

```diff
@@ -0,0 +1,304 @@
+};
+
+VNCConnectionFSM.prototype.state_upgrade = function state_upgrade(S) {
+    try {
+        this.upgrade = this.res.claimUpgrade();
+        /*
+         * Since VNC prefers low latency over high bandwidth, disable Nagle's
+         * algorithm. This means that small data packets (e.g. mouse movements
+         * or key presses) will be sent immediately instead of waiting for
+         * further data.
+         */
+        this.upgrade.socket.setNoDelay(true);
+
+        this.ws = shed.accept(this.req, this.upgrade.socket, this.upgrade.head,
+            false, ['binary', 'rfb']);
+    } catch (ex) {
+        this.log.error(ex, 'websocket upgrade failed');
+        S.gotoState('reject');
+        return;
+    }
+    /*
+     * From restify's perspective, the HTTP request ends here. We set the
+     * statusCode so that the audit logs show that we upgraded to websockets.
+     */
+    this.res.statusCode = 101;
+    this.next();
+
+    /* Now we continue on to use the websocket. */
+    S.gotoState('getport');
+};
+
+VNCConnectionFSM.prototype.state_getport = function state_getport(S) {
+    var vm = this.req.machine;
+    var uri = '/servers/' + vm.compute_node + '/vms/' + vm.id + '/vnc';
+    var self = this;
+    this.req.sdc.cnapi.get(uri, S.callback(function gotVNCDetails(err, obj) {
+        if (err) {
+            self.log.error(err, 'failed to fetch VM VNC details from CNAPI');
+            self.err = new restify.InternalServerError('Failed to retrieve ' +
+                'VNC socket details');
+            S.gotoState('error');
+            return;
+        }
+        if (typeof (obj.host) !== 'string' || typeof (obj.port) !== 'number') {
+            self.log.error({ obj: obj }, 'CNAPI returned invalid VM VNC obj');
+            self.err = new restify.InternalServerError('Failed to retrieve ' +
+                'VNC socket details');
+            S.gotoState('error');
+            return;
+        }
+        self.host = obj.host;
+        self.port = obj.port;
+        self.log = self.log.child({ vncHost: obj.host, vncPort: obj.port });
+        self.log.debug('cnapi returned address for vnc');
+        S.gotoState('connect');
+    }));
+    S.on(this.ws, 'error', function onWsError(err) {
+        S.gotoState('error');
+    });
+};
+
+VNCConnectionFSM.prototype.state_error = function state_error(S) {
+    this.log.warn(this.err, 'vnc connection exited with error');
+    if (this.ws) {
+        try {
+            this.ws.end(JSON.stringify({ type: 'error', error: this.err }));
+        } catch (ex) {
+            this.ws.destroy();
+        }
+    }
+    if (this.socket) {
+        this.socket.destroy();
+    }
+};
+
+VNCConnectionFSM.prototype.state_connect = function state_connect(S) {
+    var self = this;
+
+    S.on(this.ws, 'error', function connectWsError(err) {
+        self.err = err;
+        S.gotoState('error');
+    });
+    S.on(this.ws, 'end', function connectWsEnd() {
+        S.gotoState('error');
```

Comment:
Is this meant to go to state 'error' on ws 'end' ?

Reply:
Seems unlikely -- probably 'ws_end'